core/model/src/main/java/org/eclipse/rdf4j/model/util/GraphComparisons.java

```diff
@@ -0,0 +1,415 @@
+		if (model1 == model2) {
+			return true;
+		}
+
+		if (model1.size() != model2.size()) {
+			return false;
+		}
+
+		if (model1.contexts().size() != model2.contexts().size()) {
+			return false;
+		}
+
+		if (model1.contexts().size() > 1) {
+			// model contains more than one context (including the null context). We compare per individual context.
+			for (Resource context : model1.contexts()) {
+				if (context != null && context.isBNode()) {
+					// We currently do not handle mapping of blank nodes used as context identifiers.
+					logger.warn(
+							"isomorphism detection can not map blank nodes used as context identifiers. Comparison may give inaccurate results",
+							context
+					);
+				}
+
+				Model contextInlModel1 = model1.filter(null, null, null, context);
+				Model contextInModel2 = model2.filter(null, null, null, context);
+				if (contextInlModel1.size() != contextInModel2.size()) {
+					return false;
+				}
+				final Model canonicalizedContext1 = isoCanonicalize(contextInlModel1);
+				final Model canonicalizedContext2 = isoCanonicalize(contextInModel2);
+				if (!canonicalizedContext1.equals(canonicalizedContext2)) {
+					return false;
+				}
+			}
+			return true;
+		} else {
+			// only one context (the null context), so we're dealing with one graph only.
+			final Map<BNode, HashCode> mapping1 = getIsoCanonicalMapping(model1);
+			if (mapping1.isEmpty()) {
+				// no blank nodes in model1 - simple collection equality will do
+				return model1.equals(model2);
+			}
+			final Map<BNode, HashCode> mapping2 = getIsoCanonicalMapping(model2);
+			if (mappingsIncompatible(mapping1, mapping2)) {
+				return false;
+			}
+
+			final Model c1 = labelModel(model1, mapping1);
+			final Model c2 = labelModel(model2, mapping2);
+			return c1.equals(c2);
+		}
+	}
+
+	private static boolean mappingsIncompatible(Map<BNode, HashCode> mapping1, Map<BNode, HashCode> mapping2) {
+		if (mapping1.size() != mapping2.size()) {
+			return true;
+		}
+		Collection<HashCode> values1 = mapping1.values();
+		Collection<HashCode> values2 = mapping2.values();
+
+		if (!(values1.containsAll(values2) && values2.containsAll(values1))) {
+			return true;
+		}
+
+		return false;
+	}
+
+	protected static Model isoCanonicalize(Model m) {
+		return labelModel(m, getIsoCanonicalMapping(m));
+	}
+
+	protected static Map<BNode, HashCode> getIsoCanonicalMapping(Model m) {
+		Map<BNode, HashCode> blankNodeMapping = hashBNodes(m);
+		Multimap<HashCode, BNode> partition = partitionMapping(blankNodeMapping);
+
+		if (isFine(partition)) {
+			return blankNodeMapping;
+		}
+
+		return distinguish(m, blankNodeMapping, partition, null, new ArrayList<>(), new ArrayList<>());
+
+	}
+
+	protected static Set<BNode> getBlankNodes(Model m) {
+		final Set<BNode> blankNodes = new HashSet<>();
+
+		m.subjects().forEach(s -> {
+			if (s.isBNode()) {
+				blankNodes.add((BNode) s);
+			}
+		});
+		m.objects().forEach(o -> {
```

Comment:
Does calling m.objects and m.subjects not cause the model to do a lot of work? Wouldn't it be faster to just iterate through all the statements once and check the subject and object at the same time instead?

Reply:
Good point. I thought I had already changed that, must have accidentally rolled it back. Will fix.